README.md
## About

Low-level and minimalistic Secp256k1 bindings for Tarantool

## Installation

TODO

## Requirements

* Tarantool 1.7+
* Secp256k1

## Examples

TODO


debian/changelog
tarantool-secp256k1 (0.1.0-1) unstable; urgency=medium

  * Initial release

 -- Eugene Blikh <bigbes@gmail.com>  Fri, 18 Oct 2018 12:15:00 +0300


secp256k1/init.lua
local log      = require('log')
local digest   = require('digest')

local internal = require('secp256k1.internal')

-------------------------------------------------------------------------------

local function secp256k1_signature_tostring(self, opts)
    opts = opts or {}
    if opts.type == nil then
        opts.type = 'der'
    elseif opts.type ~= 'der' and opts.type ~= 'compact' then
        opts.type = 'der'
    end

    local rv = nil
    if opts.type == 'der' then
        rv = internal.ecdsa_signature_serialize_der(self.int)
    else
        rv = internal.ecdsa_signature_serialize_compact(self.int)
    end
    if not rv then
        log.error('Failed to serialize signature')
        return nil
    end

    return rv
end

local secp256k1_signature_mt = {
    __index = {
        tostring = secp256k1_signature_tostring,
    }
}

local function secp256k1_signature_new(input, opts)
    opts = opts or {}
    if opts.type == nil then
        opts.type = 'der'
    elseif opts.type ~= 'der' and opts.type ~= 'compact' then
        opts.type = 'der'
    end

    local self = { int = internal.int_signature_new() }
    if input then
        local rv = nil
        if opts.type == 'der' then
            rv = internal.ecdsa_signature_parse_der(self.int, input)
        else
            rv = internal.ecdsa_signature_parse_compact(self.int, input)
        end
        if not rv then
            log.error('Failed to parse signature: %s', input:hex())
            return nil
        end
    end
    return setmetatable(self, secp256k1_signature_mt)
end

-------------------------------------------------------------------------------

local function secp256k1_pubkey_verify(self, signature, hash)
    if type(signature) == 'string' then
        signature = secp256k1_signature_new(signature)
    end
    return internal.ecdsa_verify(signature.int, hash, self.int)
end

local function secp256k1_pubkey_tostring(self, opts)
    opts = opts or {}
    if opts.compressed == nil then
        opts.compressed = true
    end
    local flags = opts.compressed and 1 or 0

    local rv = internal.ec_pubkey_serialize(self.int, flags)
    if not rv then
        log.error('Failed to serialize public key')
        return nil
    end
    return rv
end

local secp256k1_pubkey_mt = {
    __index = {
        verify     = secp256k1_pubkey_verify,
        tostring   = secp256k1_pubkey_tostring,
        to_address = function(self) return self._address end,
        to_pkh     = function(self) return self._pkh     end,
    }
}

local function secp256k1_pubkey_new(pubkey, tp)
    local self = setmetatable({
        int = internal.int_pubkey_new()
    }, secp256k1_pubkey_mt)

    local rv = nil
    if #pubkey == 33 then -- public key
        rv = internal.ec_pubkey_parse(self.int, pubkey)
    elseif #pubkey == 32 then -- generate from private key
        rv = internal.ec_pubkey_create(self.int, pubkey)
    else
        error(('secp256k1.public_key_new: unknown public_key with length %s (expected 33/32)'):format(#pubkey))
    end
    if not rv then
        log.error('Failed to create pubkey')
        return nil
    end

    return self
end

-------------------------------------------------------------------------------

local function secp256k1_privkey_sign(self, hash)
    local signature = secp256k1_signature_new()

    local rv = internal.ecdsa_sign(signature.int, hash, self.privkey)
    if not rv then
        return nil
    end

    return signature
end

local secp256k1_privkey_mt = {
    __index = {
        sign     = secp256k1_privkey_sign,
        tostring = function(self)      return self.privkey            end,
        verify   = function(self, ...) return self.pubkey:verify(...) end,
    }
}

local function secp256k1_privkey_new(privkey)
    local self = { }
    if privkey == nil then -- generate private key
        repeat
            privkey = digest.urandom(32)
        until internal.ec_seckey_verify(privkey)
    end
    self.privkey = privkey
    self.pubkey  = secp256k1_pubkey_new(privkey)
    return setmetatable(self, secp256k1_privkey_mt)
end

-------------------------------------------------------------------------------

return {
    signature     = secp256k1_signature_new,
    public_key    = secp256k1_pubkey_new,
    private_key   = secp256k1_privkey_new,
}


secp256k1/internal.c
#include <module.h> /* <tarantool/module.h> */

#include <errno.h>
#include <assert.h>
#include <limits.h>
#include <stdarg.h>

#include <lua.h>
#include <lauxlib.h>
#include <lualib.h>

#include <secp256k1.h>

static secp256k1_context *context;
static secp256k1_context *sign_context;
static secp256k1_context *verify_context;

static const char pubkey_typename[]    = "secp256k1.pubkey";
static const char signature_typename[] = "secp256k1.signature";

LUA_API int
Tsecp256k1_int_signature_new(lua_State *L)
{
	void *p = lua_newuserdata(L, sizeof(secp256k1_ecdsa_signature) * 2);
	memset(p, 0, sizeof(secp256k1_ecdsa_signature));
	luaL_getmetatable(L, signature_typename);
	lua_setmetatable(L, -2);
	return 1;
}

LUA_API int
Tsecp256k1_int_pubkey_new(lua_State *L)
{
	void *p = lua_newuserdata(L, sizeof(secp256k1_pubkey) * 2);
	memset(p, 0, sizeof(secp256k1_pubkey));
	luaL_getmetatable(L, pubkey_typename);
	lua_setmetatable(L, -2);
	return 1;
}

/* 
 * <- pubkey (string), pubkey (udata)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ec_pubkey_parse(lua_State *L)
{
	secp256k1_pubkey *pubkey_obj = luaL_checkudata(L, 1, pubkey_typename);
	size_t pubkey_len = 0;
	const unsigned char *pubkey;
	pubkey = (const unsigned char *)luaL_checklstring(L, 2, &pubkey_len);

	int rv = secp256k1_ec_pubkey_parse(context, pubkey_obj,
					   pubkey, pubkey_len);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- pubkey (udata), is_compressed (int, opt == 1)
 * -> string / false
 */
LUA_API int
Tsecp256k1_ec_pubkey_serialize(lua_State *L) {
	secp256k1_pubkey *pubkey_obj = luaL_checkudata(L, 1, pubkey_typename);

	int flags = SECP256K1_EC_COMPRESSED;
	if (luaL_optnumber(L, 2, 1) == 0) {
		flags = SECP256K1_EC_UNCOMPRESSED;
	}

	unsigned char pubkey[66]; size_t pubkey_len = 66;
	int rv = secp256k1_ec_pubkey_serialize(context,
			(unsigned char *)pubkey, &pubkey_len, pubkey_obj, flags);
	if (rv == 1) {
		lua_pushlstring(L, (const char *)pubkey, pubkey_len);
		return 1;
	}
	return 0;
}

/*
 * <- signature (udata), input (string)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ecdsa_signature_parse_compact(lua_State *L) {
	secp256k1_ecdsa_signature *signature_obj;
	signature_obj = luaL_checkudata(L, 1, signature_typename);

	size_t signature_len = 0;
	const unsigned char *signature;
	signature = (const unsigned char *)luaL_checklstring(L, 2, &signature_len);
	assert(signature_len == 64);

	int rv = secp256k1_ecdsa_signature_parse_compact(context, signature_obj,
							 signature);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- signature (udata), input (string)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ecdsa_signature_parse_der(lua_State *L) {
	secp256k1_ecdsa_signature *signature_obj;
	signature_obj = luaL_checkudata(L, 1, signature_typename);

	size_t signature_len = 0;
	const unsigned char *signature;
	signature = (const unsigned char *)luaL_checklstring(L, 2, &signature_len);

	int rv = secp256k1_ecdsa_signature_parse_der(context, signature_obj,
						     signature, signature_len);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- signature (udata)
 * -> string / false
 */
LUA_API int
Tsecp256k1_ecdsa_signature_serialize_der(lua_State *L) {
	secp256k1_ecdsa_signature *signature_obj;
	signature_obj = luaL_checkudata(L, 1, signature_typename);

	size_t signature_len = 128; unsigned char signature[128];

	int rv = secp256k1_ecdsa_signature_serialize_der(
			context, signature, &signature_len, signature_obj);

	if (rv == 1) {
		lua_pushlstring(L, (const char *)signature, signature_len);
		return 1;
	}
	return 0;
}

/*
 * <- signature (udata)
 * -> string / false
 */
LUA_API int
Tsecp256k1_ecdsa_signature_serialize_compact(lua_State *L) {
	secp256k1_ecdsa_signature *signature_obj;
	signature_obj = luaL_checkudata(L, 1, signature_typename);

	size_t signature_len = 64;
	unsigned char signature[64];

	int rv = secp256k1_ecdsa_signature_serialize_compact(
			context, signature, signature_obj);

	if (rv == 1) {
		lua_pushlstring(L, (const char *)signature, signature_len);
		return 1;
	}
	return 0;
}

/*
 * <- signature (udata), hash(string, 32bytes), pubkey (udata)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ecdsa_verify(lua_State *L) {
	secp256k1_pubkey *pubkey_obj;
	secp256k1_ecdsa_signature *signature_obj;
	size_t hash32_len; const unsigned char *hash32;

	signature_obj = luaL_checkudata(L, 1, signature_typename);
	hash32 = (const unsigned char *)luaL_checklstring(L, 2, &hash32_len);
	pubkey_obj = luaL_checkudata(L, 3, pubkey_typename);

	int rv = secp256k1_ecdsa_verify(verify_context, signature_obj, hash32,
					pubkey_obj);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- signature (udata), hash(string, 32bytes), seckey
 * -> true / false
 */
LUA_API int
Tsecp256k1_ecdsa_sign(lua_State *L) {
	secp256k1_ecdsa_signature *signature_obj;
	size_t privkey_len, hash32_len; const unsigned char *hash32, *privkey;

	signature_obj = luaL_checkudata(L, 1, signature_typename);
	hash32 = (const unsigned char *)luaL_checklstring(L, 2, &hash32_len);
	privkey = (const unsigned char *)luaL_checklstring(L, 3, &privkey_len);

	int rv = secp256k1_ecdsa_sign(sign_context, signature_obj, hash32,
				      privkey, NULL, NULL);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- private key(string, 32 bytes)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ec_seckey_verify(lua_State *L) {
	size_t privkey_len; const unsigned char *privkey;
	privkey = (const unsigned char *)luaL_checklstring(L, 1, &privkey_len);

	int rv = secp256k1_ec_seckey_verify(sign_context, privkey);

	lua_pushboolean(L, rv);
	return 1;
}

/*
 * <- public (udata), private(string, 32 bytes)
 * -> true / false
 */
LUA_API int
Tsecp256k1_ec_pubkey_create(lua_State *L) {
	secp256k1_pubkey *pubkey_obj;
	size_t privkey_len; const unsigned char *privkey;

	pubkey_obj = luaL_checkudata(L, 1, pubkey_typename);
	privkey = (const unsigned char *)luaL_checklstring(L, 2, &privkey_len);

	int rv = secp256k1_ec_pubkey_create(sign_context, pubkey_obj, privkey);

	lua_pushboolean(L, rv);
	return 1;
}

void secp256k1_handler(const char *message, void *data) {
	(void )data;
	say_error(message);
}

LUA_API int
luaopen_secp256k1_internal(lua_State *L)
{

	context        = secp256k1_context_create(SECP256K1_CONTEXT_NONE);
	sign_context   = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
	verify_context = secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
	assert(context && sign_context && verify_context);

	#if 0
	secp256k1_context_set_error_callback(context,          &secp256k1_handler, NULL);
	secp256k1_context_set_error_callback(sign_context,     &secp256k1_handler, NULL);
	secp256k1_context_set_error_callback(verify_context,   &secp256k1_handler, NULL);
	secp256k1_context_set_illegal_callback(context,        &secp256k1_handler, NULL);
	secp256k1_context_set_illegal_callback(sign_context,   &secp256k1_handler, NULL);
	secp256k1_context_set_illegal_callback(verify_context, &secp256k1_handler, NULL);
	#endif

	static const struct luaL_Reg signature_meta[] = {
		{NULL, NULL}
	};
	luaL_newmetatable(L, signature_typename);
	lua_pushstring(L, signature_typename);
	lua_setfield(L, -2, "__metatable");
	luaL_register(L, NULL, signature_meta);
	lua_pop(L, 1);

	static const struct luaL_Reg pubkey_meta[] = {
		{NULL, NULL}
	};
	luaL_newmetatable(L, pubkey_typename);
	lua_pushstring(L, pubkey_typename);
	lua_setfield(L, -2, "__metatable");
	luaL_register(L, NULL, pubkey_meta);
	lua_pop(L, 1);

	lua_newtable(L);
	static const struct luaL_Reg funcs[] = {
		{ "int_signature_new",                 Tsecp256k1_int_signature_new                 },
		{ "int_pubkey_new",                    Tsecp256k1_int_pubkey_new                    },
		{ "ec_pubkey_parse",                   Tsecp256k1_ec_pubkey_parse                   },
		{ "ec_pubkey_serialize",               Tsecp256k1_ec_pubkey_serialize               },
		{ "ecdsa_signature_parse_compact",     Tsecp256k1_ecdsa_signature_parse_compact     },
		{ "ecdsa_signature_parse_der",         Tsecp256k1_ecdsa_signature_parse_der         },
		{ "ecdsa_signature_serialize_der",     Tsecp256k1_ecdsa_signature_serialize_der     },
		{ "ecdsa_signature_serialize_compact", Tsecp256k1_ecdsa_signature_serialize_compact },
		{ "ecdsa_verify",                      Tsecp256k1_ecdsa_verify                      },
		{ "ecdsa_sign",                        Tsecp256k1_ecdsa_sign                        },
		{ "ec_seckey_verify",                  Tsecp256k1_ec_seckey_verify                  },
		{ "ec_pubkey_create",                  Tsecp256k1_ec_pubkey_create                  },
		{ NULL,                                NULL                                         }
	};
	luaL_register(L, NULL, funcs);

	return 1;
}


secp256k1/init.ffi.old.lua
local ffi    = require('ffi')
local log    = require('log')
local digest = require('digest')

local ibuf   = require('buffer').IBUF_SHARED

ffi.cdef[[
typedef struct {
    unsigned char data[64];
} secp256k1_pubkey;

typedef struct {
    unsigned char data[64];
} secp256k1_ecdsa_signature;

typedef int (*secp256k1_nonce_function)(
    unsigned char *nonce32,
    const unsigned char *msg32,
    const unsigned char *key32,
    const unsigned char *algo16,
    void *data,
    unsigned int attempt
);

/* Context and context creation */
struct secp256k1_context_struct;
typedef struct secp256k1_context_struct secp256k1_context;

secp256k1_context* secp256k1_context_create(unsigned int flags);
void secp256k1_context_destroy(secp256k1_context* ctx);

int secp256k1_ec_pubkey_parse(
    const secp256k1_context* ctx,
    secp256k1_pubkey* pubkey,
    const unsigned char *input,
    size_t inputlen
);

int secp256k1_ec_pubkey_serialize(
    const secp256k1_context* ctx,
    unsigned char *output,
    size_t *outputlen,
    const secp256k1_pubkey* pubkey,
    unsigned int flags
);

int secp256k1_ecdsa_signature_parse_compact(
    const secp256k1_context* ctx,
    secp256k1_ecdsa_signature* sig,
    const unsigned char *input64
);

int secp256k1_ecdsa_signature_serialize_compact(
    const secp256k1_context* ctx,
    unsigned char *output64,
    const secp256k1_ecdsa_signature* sig
);

int secp256k1_ecdsa_signature_parse_der(
    const secp256k1_context* ctx,
    secp256k1_ecdsa_signature* sig,
    const unsigned char *input,
    size_t inputlen
);

int secp256k1_ecdsa_signature_serialize_der(
    const secp256k1_context* ctx,
    unsigned char *output,
    size_t *outputlen,
    const secp256k1_ecdsa_signature* sig
);

int secp256k1_ecdsa_verify(
    const secp256k1_context* ctx,
    const secp256k1_ecdsa_signature *sig,
    const unsigned char *msg32,
    const secp256k1_pubkey *pubkey
);

int secp256k1_ecdsa_sign(
    const secp256k1_context* ctx,
    secp256k1_ecdsa_signature *sig,
    const unsigned char *msg32,
    const unsigned char *seckey,
    secp256k1_nonce_function noncefp,
    const void *ndata
);

int secp256k1_ec_seckey_verify(
    const secp256k1_context* ctx,
    const unsigned char *seckey
);

int secp256k1_ec_pubkey_create(
    const secp256k1_context* ctx,
    secp256k1_pubkey *pubkey,
    const unsigned char *seckey
);

int secp256k1_context_randomize(
    secp256k1_context* ctx,
    const unsigned char *seed32
);
]]

local secp256k1 = ffi.load('secp256k1')

local SECP256K1_FLAGS_TYPE_MASK          = bit.lshift(1, 8) - 1
local SECP256K1_FLAGS_TYPE_CONTEXT       = bit.lshift(1, 0)
local SECP256K1_FLAGS_TYPE_COMPRESSION   = bit.lshift(1, 1)
local SECP256K1_FLAGS_BIT_CONTEXT_VERIFY = bit.lshift(1, 8)
local SECP256K1_FLAGS_BIT_CONTEXT_SIGN   = bit.lshift(1, 9)
local SECP256K1_FLAGS_BIT_COMPRESSION    = bit.lshift(1, 8)

local SECP256K1_CONTEXT_VERIFY = bit.bor(SECP256K1_FLAGS_TYPE_CONTEXT,
                                         SECP256K1_FLAGS_BIT_CONTEXT_VERIFY)
local SECP256K1_CONTEXT_SIGN   = bit.bor(SECP256K1_FLAGS_TYPE_CONTEXT,
                                         SECP256K1_FLAGS_BIT_CONTEXT_SIGN)
local SECP256K1_CONTEXT_NONE   = bit.bor(SECP256K1_FLAGS_TYPE_CONTEXT)

local SECP256K1_EC_COMPRESSED   = bit.bor(SECP256K1_FLAGS_TYPE_COMPRESSION,
                                          SECP256K1_FLAGS_BIT_COMPRESSION)
local SECP256K1_EC_UNCOMPRESSED = bit.bor(SECP256K1_FLAGS_TYPE_COMPRESSION)

-------------------------------------------------------------------------------

local function secp256k1_gen_ctx(flags)
    local ctx = secp256k1.secp256k1_context_create(flags)
    assert(ctx, 'cp256k1_context_create failed')
    return ffi.gc(ctx, function(obj) log.info('destroyed'); secp256k1.secp256k1_context_destroy(obj) end)
end

local function secp256k1_ctx_verify(self, key)
    assert(#key == 32)
    return (secp256k1.secp256k1_ec_seckey_verify(self.int, key) == 1)
end

local secp256k1_ctx_mt = {
    __index = {
        verify = secp256k1_ctx_verify
    }
}

local function secp256k1_ctx_new(flags)
    local int = secp256k1_gen_ctx(flags)

    --[[
    if bit.band(flags, SECP256K1_FLAGS_BIT_CONTEXT_SIGN) ~= 0 then
        local seed = digest.urandom(32)
        local rv = secp256k1.secp256k1_context_randomize(int, seed)
        assert(rv == 1, 'cp256k1_context_randomize failed to execute')
    end
    ]]--

    return setmetatable({ int = int }, secp256k1_ctx_mt)
end

-------------------------------------------------------------------------------

local ctx        = secp256k1_ctx_new(SECP256K1_CONTEXT_NONE  )
local sign_ctx   = secp256k1_ctx_new(SECP256K1_CONTEXT_SIGN  )
local verify_ctx = secp256k1_ctx_new(SECP256K1_CONTEXT_VERIFY)

local static_size_t_buf = ffi.new('size_t [1]', 0)

-------------------------------------------------------------------------------

local function secp256k1_gen_signature()
    local signature_arr = ffi.new('secp256k1_ecdsa_signature [1]')
    assert(signature_arr)
    ffi.fill(signature_arr[1].data, 64, 0)
    return signature_arr
end

local function secp256k1_signature_tostring(self, opts)
    opts = opts or {}
    if opts.type == nil then
        opts.type = 'der'
    elseif opts.type ~= 'der' and opts.type ~= 'compact' then
        opts.type = 'der'
    end

    ibuf:recycle()
    static_size_t_buf[0] = 256
    local bpos = ibuf:alloc(256)

    local rv = nil
    if opts.type == 'der' then
        rv = secp256k1.secp256k1_ecdsa_signature_serialize_der(
            sign_ctx.int, bpos, static_size_t_buf, self.int
        )
    else
        rv = secp256k1.secp256k1_ecdsa_signature_serialize_compact(
            sign_ctx.int, bpos, self.int
        )
        static_size_t_buf[0] = 64
    end
    if rv == -1 then
        log.error('Failed to serialize signature')
        return nil
    end

    return ffi.string(bpos, tonumber(static_size_t_buf[0]))
end

local secp256k1_signature_mt = {
    __index = {
        tostring = secp256k1_signature_tostring,
    }
}

local function secp256k1_signature_new(input, opts)
    opts = opts or {}
    if opts.type == nil then
        opts.type = 'der'
    elseif opts.type ~= 'der' and opts.type ~= 'compact' then
        opts.type = 'der'
    end

    local self = {
        int = secp256k1_gen_signature()
    }
    if input then
        local rv = nil
        if opts.type == 'der' then
            rv = secp256k1.secp256k1_ecdsa_signature_parse_der(
                sign_ctx.int, self.int, input, #input
            )
        else
            rv = secp256k1.secp256k1_ecdsa_signature_parse_compact(
                sign_ctx.int, self.int, input
            )
        end
        if rv ~= 1 then
            log.error('Failed to parse compact signature: %s', input:hex())
            return nil
        end
    end
    return setmetatable(self, secp256k1_signature_mt)
end

-------------------------------------------------------------------------------

local function secp256k1_pubkey_verify(self, signature, hash)
    if type(signature) == 'str' then
        signature = secp256k1_signature_new(signature)
    end
    local rv = secp256k1.secp256k1_ecdsa_verify(
        verify_ctx.int, signature.int, hash, self.int
    )
    return rv == 1
end

local function secp256k1_pubkey_tostring(self, opts)
    opts = opts or {}
    if opts.compressed == nil then
        opts.compressed = true
    end
    local flags = opts.compressed and SECP256K1_EC_COMPRESSED
                                   or SECP256K1_EC_UNCOMPRESSED

    ibuf:recycle()
    static_size_t_buf[0] = 128;
    local bpos = ibuf:alloc(128)
    local rv = secp256k1.secp256k1_ec_pubkey_serialize(
        sign_ctx.int, bpos, static_size_t_buf, self.int, flags
    )
    if rv == -1 then
        log.error('Failed to serialize public key')
        return nil
    end
    return ffi.string(bpos, tonumber(static_size_t_buf[0]))
end

local secp256k1_pubkey_mt = {
    __index = {
        verify     = secp256k1_pubkey_verify,
        tostring   = secp256k1_pubkey_tostring,
        to_address = function(self) return self._address end,
        to_pkh     = function(self) return self._pkh     end,
    }
}

local function secp256k1_gen_pubkey()
    local pubkey_arr = ffi.new('secp256k1_pubkey [1]')
    ffi.fill(pubkey_arr[1].data, 64, 0)
    return pubkey_arr
end

local function secp256k1_pubkey_new(pubkey, tp)
    local self = setmetatable({
        int = secp256k1_gen_pubkey(),
    }, secp256k1_pubkey_mt)

    if #pubkey == 33 then -- public key
        local rv = secp256k1.secp256k1_ec_pubkey_parse(
            sign_ctx.int, self.int, pubkey, #pubkey
        )
        if rv ~= 1 then
            return nil
        end
    elseif #pubkey == 32 then -- generate from private key
        local rv = secp256k1.secp256k1_ec_pubkey_create(
            sign_ctx.int, self.int, pubkey
        )
        if rv ~= 1 then
            return nil
        end
    else
        assert(false)
    end

    return self
end

-------------------------------------------------------------------------------

local function secp256k1_privkey_sign(self, hash)
    local signature = secp256k1_signature_new()

    local rv = secp256k1.secp256k1_ecdsa_sign(
        sign_ctx.int, signature.int, hash, self.privkey, nil, nil
    )
    if rv ~= 1 then
        return nil
    end

    return signature
end

local secp256k1_privkey_mt = {
    __index = {
        sign     = secp256k1_privkey_sign,
        tostring = function(self)      return self.privkey            end,
        verify   = function(self, ...) return self.pubkey:verify(...) end,
    }
}

local function secp256k1_privkey_new(privkey)
    local self = { }
    if privkey == nil then -- generate private key
        repeat
            privkey = digest.urandom(32)
        until sign_ctx:verify(privkey)
    end
    self.privkey = privkey
    self.pubkey  = secp256k1_pubkey_new(privkey)
    return setmetatable(self, secp256k1_privkey_mt)
end

-------------------------------------------------------------------------------

return {
    signature     = secp256k1_signature_new,
    public_key    = secp256k1_pubkey_new,
    private_key   = secp256k1_privkey_new,
}
